src/wayback/_http.py

```diff
@@ -0,0 +1,591 @@
+            # gzip/brotli/etc. responses, so we need to handle those errors by
+            # continuing to just read the raw data off the socket instead.
+            #
+            # This fallback behavior is inspired by requests:
+            #   https://github.com/psf/requests/blob/eedd67462819f8dbf8c1c32e77f9070606605231/requests/sessions.py#L160-L163
+            # For urllib3, the appropriate errors to handle would be:
+            #   `(DecodeError, ProtocolError, RuntimeError)`
+            try:
+                if cache:
+                    try:
+                        self.content
+                    except (ChunkedEncodingError, ContentDecodingError, RuntimeError):
+                        self._raw.raw.read(decode_content=False)
+                else:
+                    self._raw.raw.read(decode_content=False)
+            finally:
+                self._raw.close()
+
+
+class WaybackHttpAdapter:
+    """
+    Handles making actual HTTP requests over the network. This is an abstract
+    base class that defines the API an adapter must implement.
+    """
+
+    def request(
+        self,
+        method: str,
+        url: str,
+        *,
+        params: dict = None,
+        headers: dict = None,
+        follow_redirects: bool = True,
+        timeout: Union[int, Tuple[int, int]] = None
+    ) -> WaybackHttpResponse:
+        """
+        Send an HTTP request and return a :class:`WaybackHttpResponse` object
+        representing the response.
+
+        Parameters
+        ----------
+        method : str
+            Method to use for the request, e.g. ``'GET'``, ``'POST'``.
+        url : str
+            The URL to reqeust.
+        params : dict, optional
+            For POST/PUT requests, data to be encoded in the response body. For
+            other methods, this should be encoded as the querystring.
+        headers : dict, optional
+            The HTTP headers to send with the request.
+        follow_redirects : bool, default: True
+            Whether to follow redirects before returning a response.
+        timeout : int or float or tuple of (int or float, int or float), optional
+            How long to wait, in seconds, before timing out. If this is a single
+            number, it will be used as both a connect timeout (how long to wait
+            for the first bit of response data) and a read timeout (how long
+            to wait between bits of response data). If a tuple, the values will
+            be used as the connect and read timeouts, respectively.
+
+        Returns
+        -------
+        WaybackHttpResponse
+        """
+        raise NotImplementedError()
+
+    def close(self) -> None:
+        """
+        Close the adapter and release any long-lived resources, like pooled
+        HTTP connections.
+        """
+        ...
+
+
+class RequestsAdapter(WaybackHttpAdapter):
+    """
+    Wrap the requests library for making HTTP requests.
+    """
```

Comment:
We now have two similarly named classes, which might be kind of dangerous:

- `RequestsAdapter` is private and just implements HTTP requests with the *requests* library. I wonder if it should have an underscore prefix or be renamed in some bigger way for clarity. (I *do* think keeping it separate is good, since it will get way more complex when we add thread safety in the next PR.)

- `WaybackRequestsAdapter` adds retry, rate limiting, etc. to the above class. It is the actual adapter that gets used. I’m also wondering if this should be renamed to something more generic or have a generic alias like `WaybackDefaultAdapter`. That way the public name is something people can rely on, so they don’t break if/when we swap out requests for raw urllib3 or httpx.